NSEC3 and multiple key signing support.

ximon18 opened the pull request:
Currently lacks collision detection and tests, though has been manually tested using `ldns-verify-zone`, `dnssec-verify` and `named-checkzone` both with and without opt-out and also including both signed and unsigned delegations.

I'm posting this here as a draft to allow for alignment and early feedback from the team working on various pieces of  DNSSEC support for `domain`.

Note: This PR includes the content of https://github.com/NLnetLabs/domain/pull/444.

ximon18 commented on src/sign/ring.rs:
```diff
@@ -207,6 +213,121 @@ impl<'a> Sign for SecretKey<'a> {
+/// iterations as defined in [RFC 5155] section 3.1.3.
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_hash<N, SaltOcts, HashOcts>(
+    owner: N,
+    algorithm: Nsec3HashAlg,
+    iterations: u16,
+    salt: &Nsec3Salt<SaltOcts>,
+) -> Result<OwnerHash<HashOcts>, Nsec3HashError>
```
> Why not just take the owner name and an `Nsec3param`?

I had that but took it out. There was a reason. I'll see if I can remember why.

ximon18 replied:
Ah I know why. The `nsec3s()` function takes an `Nsec3Param` struct which it uses to create the `NSEC3PARAM` RR at the apex of the zone. For the `NSEC3PARAM` RR RFC 5155 says that the opt-out flag in the flags field MUST be zero, so for RFC compliance the `Nsec3Param` struct passed to `nsec3s()` should have the opt-out flag set to zero.

Honestly I think this is a bit of a foot-gun and perhaps best not to pass an `Nsec3Param` to `nsec3s()` but instead only the other fields (algorithm, iterations, salt), however MAYBE in future it will be legal to set some of the other flag bits in the flags field and a user would want to have those set in the created `NSEC3PARAM` RR... so for that reason `nsec3s()` currently takes an `Nsec3Param` as input.

When generating NSEC3 RRs and when opt-out is enabled the flags value in the given `Nsec3Param` cannot be used as-is because the opt-out flag must be set to 1 (but NOT in the `NSEC3PARAM` RR), ad rather than copy the given `Nsec3Param` or modify it and then pass it to `nsec3_hash()` I felt it was better to just pass only the values actually needed for hashing in, as NSEC3 hashing doesn't need the flags field at all, and also that way users don't have to think about what value to set the unused `Nsec3Param::flags` field to when invoking `nsec3_hash()` directly (as `dnst nsec3-hash` does).

ximon18 commented on src/sign/ring.rs:
```diff
@@ -207,6 +213,121 @@ impl<'a> Sign for SecretKey<'a> {
+    fn mk_hash<N, SaltOcts, HashOcts>(
+        owner: N,
+        iterations: u16,
+        salt: &Nsec3Salt<SaltOcts>,
```
I don't know if I agree or not, it would be useful to know your rationale for this opinion.

bal-e replied:
Yeah, we don't need to worry about it right now :)

ximon18 commented on src/sign/records.rs:
```diff
@@ -299,6 +621,29 @@ where
+
+impl<N, Octets> Nsec3Records<N, Octets> {
+    pub fn new(
+        nsec3_recs: Vec<Record<N, Nsec3<Octets>>>,
+        nsec3param_rec: Record<N, Nsec3param<Octets>>,
```
I was expecting to revisit the design of these APIs. If it remains as such then yes I agree.

ximon18 commented on src/sign/ring.rs:
```diff
@@ -351,6 +353,121 @@ impl fmt::Display for GenerateError {
+        let mut buf = HashOcts::empty();
+
+        owner.compose_canonical(&mut buf)?;
+        buf.append_slice(salt.as_slice())?;
```
I suspect this code was already so before I touched it, will have to check.

ximon18 replied:
`owner.compose_canonical()` needs a buffer to write into so that requires `buf` but the salt could be added to the context directly.

bal-e replied:
Oh, right ... that's unfortunate.  We could replace the `N` parameter with `Name<Octs>` instead (which is likely what it will receive anyway), but I'll probably end up addressing that with the `new_name` module.

ximon18 commented on src/sign/ring.rs:
```diff
@@ -207,6 +213,121 @@ impl<'a> Sign for SecretKey<'a> {
+    let hash = mk_hash(owner, iterations, salt)
+        .map_err(|_| Nsec3HashError::AppendError)?;
+
+    let owner_hash = OwnerHash::from_octets(hash)
+        .map_err(|_| Nsec3HashError::OwnerHashError)?;
```
It's not clear to me from the docs on `OwnerHashError` that it only fails relating to length, it just says "The hashing process produced an invalid owner hash" and I have no way of knowing when that error might occur or why.

ximon18 commented:
@bal-e: I realize that I moved the `nsec3_hash()` function from `validator/nsec.rs` to `sign/ring.rs` I think because that more clearly showed that it was Ring powered/dependent. However, we were discussing yesterday that lots of functionality is shared between signing and validating and that right now you are putting everything under validation, while I proposed a common module shared by both signing and validating. Either way, if you are currently putting everything under validation, this move to sign is inconsistent and perhaps you think I should move it back?

bal-e commented:
@ximon18: yeah, I think this should be moved under `validate` for now.  We'll gather up the shared functionality in that module for now, then decide how to distribute it.

ximon18 commented:
> @ximon18: yeah, I think this should be moved under `validate` for now. We'll gather up the shared functionality in that module for now, then decide how to distribute it.

Done.

bal-e commented on src/validate.rs:
```diff
@@ -1682,3 +1685,117 @@ mod test {
+        let mut h = ctx.finish();
+
+        for _ in 0..iterations {
+            canonical_owner.truncate(0);
+            canonical_owner.append_slice(h.as_ref())?;
+            canonical_owner.append_slice(salt.as_slice())?;
```
The `canonical_owner` buffer is unnecessary here.  Directly append these slices into the digest context.

ximon18 commented on src/sign/records.rs:
```diff
@@ -243,6 +256,239 @@ impl<N, D> SortedRecords<N, D> {
+            //    the apex and the original owner name."
+            let distance_to_root = name.owner().iter_labels().count();
+            let distance_to_apex = distance_to_root - apex_label_count;
+            if distance_to_apex > 1 {
```
True, but the if statement matches nicely with the RFC text and makes it clear that if we enter this block it is because of the condition identified by the RFC text.

ximon18 commented on src/sign/records.rs:
```diff
@@ -243,6 +256,239 @@ impl<N, D> SortedRecords<N, D> {
+                // It will NOT construct the last name as that will be dealt
+                // with in the next outer loop iteration.
+                //   - a.b.c.mail.example.com
+                for n in (1..distance_to_apex - 1).rev() {
```
Thanks! I was planning to add proper test cases but for now have extended my test input zone to cover this missing case and realize now why I had some additional logic in there before which I removed because I couldn't see what value it was adding... 😛

ximon18 commented on src/sign/records.rs:
```diff
@@ -243,6 +256,239 @@ impl<N, D> SortedRecords<N, D> {
+    ///   SOA RR and the TTL of the zone SOA RR itself"_.
+    ///
+    /// - The `params` should be set to _"SHA-1, no extra iterations, empty
+    ///   salt"_ and zero flags. See `Nsec3param::default()`.
```
If you mean `Nsec3param::default()` - yes, will fix.

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Merge branch 'initial-nsec3-generation' into multiple-key-signing

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Make key selection more flexible. (#464)

E.g. LDNS seems to consider DS and CDS and CDNSKEY resource record types as well as DNSKEY when selecting keys.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Add a tab before the RDATA as well as within it, to match LDNS tabbed output format. (#463)

ximon18 committed:
Add a tab before the RDATA as well as within it, to match LDNS tabbed output format. (#463)

ximon18 committed:
Add a tab before the RDATA as well as within it, to match LDNS tabbed output format. (#463)

ximon18 committed:
Add a tab before the RDATA as well as within it, to match LDNS tabbed output format. (#463)

ximon18 committed:
Update changelog.

ximon18 committed:
Update changelog.

ximon18 committed:
Update changelog.

ximon18 committed:
Update changelog.

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Raise errors instead of unwrapping on missing apex.

ximon18 committed:
Add a logging related TODO.

ximon18 committed:
Also log the key tag when debug logging the keys to use for signing.

ximon18 committed:
Don't emit duplicate DNSKEY RRs for zonefiles that already contain the DNSKEY that matches one of the keys to sign the zone with.

Also, only sign DNSKEY RRs with key signing keys if the DNSKEY RR is at the apex.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
FIX: When extending SortedRecords, don't permit duplicate RRs to creep in, as insert() would have prevented these.

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Don't attempt to sign a zone or select keys to use if no keys are provided or found to be suitable.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Improve signing keys debug output.

ximon18 committed:
FIX: Only sign the apex if given the apex and remove unnecessary error variants.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Actually check that we were given THE apex, not AN apex.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Extend zone parsing to let the caller know when the origin has been detected and with what value, so that the caller can reliably ignore apex records if needed (e.g. when updating NSEC3PARAM or ZONEMD RRs) when signing a loaded zone. This is a breaking change.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Revert "Extend zone parsing to let the caller know when the origin has been detected and with what value, so that the caller can reliably ignore apex records if needed (e.g. when updating NSEC3PARAM or ZONEMD RRs) when signing a loaded zone. This is a breaking change."

This reverts commit 605efe6148695c4d8d2cd8840b82adb57f89f37c.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Use the correct TTL for added DNSKEY RRs when signing.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
FIX: Don't allow duplicate RRs to be imported via `impl From`.

ximon18 committed:
Add a comment explaining why the apex name we use for an RRSIG meets the RFC requirements for it to be canonical and uncompressed.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Merge branch 'initial-nsec3-generation' into multiple-key-signing

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
FIX: Sign a merged DNSKEY RR set containing existing and new DNSKEY RRs, not two separate DNSKEY RRsets (existing and new).

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

Philip-NLnetLabs commented on src/sign/records.rs:
```diff
@@ -149,31 +163,34 @@ impl<N, D> SortedRecords<N, D> {
+                let signature = signature.as_ref().to_vec();
+                let Ok(signature) = signature.try_octets_into() else {
+                    return Err(ErrorTypeToBeDetermined);
+                };
```
I think that signing an RRset should be a separated function and not be inlined here.

Philip-NLnetLabs commented on src/sign/records.rs:
```diff
@@ -609,3 +1046,54 @@ where
+//------------ ErrorTypeToBeDetermined ---------------------------------------
+
+#[derive(Debug)]
+pub struct ErrorTypeToBeDetermined;
```
I guess this needs to be renamed

Philip-NLnetLabs commented on src/validate.rs:
```diff
@@ -1735,3 +1738,138 @@ mod test {
+/// iterations as defined in [RFC 5155] section 3.1.3.
+///
+/// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155
+pub fn nsec3_hash<N, SaltOcts, HashOcts>(
```
I think we need to avoid adding new stuff to src/validate.rs. Certainly not stuff that is needed for signing.

Philip-NLnetLabs commented on src/validator/nsec.rs:
```diff
@@ -1,22 +1,25 @@
+use std::str::{FromStr, Utf8Error};
+use std::sync::Arc;
+use std::vec::Vec;
+
```
Why are there new empty lines in the use statements?

ximon18 replied:
The imports were re-ordered and grouped to match https://github.com/NLnetLabs/domain/blob/contributing-doc/CONTRIBUTING.md#imports which is not strictly related to this PR so could be skipped, but presumably I did it siince I was touching the file anyway.

Philip-NLnetLabs replied:
It says: "Imports are not automatically reordered and must be sorted manually. We sort them alphabetically in a series of groups: core, std, extern crates, crate, super. There are no empty lines between groups."

This is also wrong. rustfmt does sort use statement.

Philip-NLnetLabs commented on src/sign/records.rs:
```diff
@@ -247,20 +268,427 @@ impl<N, D> SortedRecords<N, D> {
+        // we can skip everything before that.
+        families.skip_before(apex);
+
+        // We also need the apex for the last NSEC.
```
```suggestion
        // We also need the apex for the last NSEC3.
```

Philip-NLnetLabs commented on src/sign/records.rs:
```diff
@@ -247,20 +268,427 @@ impl<N, D> SortedRecords<N, D> {
+                //   - b.c.mail.example.com
+                // It will NOT construct the last name as that will be dealt
+                // with in the next outer loop iteration.
+                //   - a.b.c.mail.example.com
```
Suppose a.mail.example.com was the last one. Would this still be correct?

Philip-NLnetLabs commented on src/sign/records.rs:
```diff
@@ -247,20 +268,427 @@ impl<N, D> SortedRecords<N, D> {
+            let mut bitmap = RtypeBitmap::<Octets>::builder();
+
+            // Authoritative RRsets will be signed.
+            if cut.is_none() || has_ds {
```
Please comment on why this test does the right thing.

partim committed:
Clippy-suggested code improvements.

ximon18 committed:
Don't hard-code NSEC3 hash capture, instead use a HashProvider.

ximon18 committed:
WIP: Use a hash provider.

partim committed:
Update changelog.

ximon18 committed:
[zonefile/inplace] Handle backslashes in quoted strings

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Merge branch 'initial-nsec3-generation' into multiple-key-signing

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
FIX: Don't omit DNSKEY RRs when signing if there were no pre-exisitng DNSKEY RRs.

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Merge branch 'multiple-key-signing' into byo-signing-sorter-impl

ximon18 committed:
Merge branch 'byo-signing-sorter-impl' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace' into sortedrecords-zonemd-remove-replace-plus-pr444

ximon18 committed:
Cargo fmt.

ximon18 committed:
Merge branch 'sortedrecords-zonemd-remove-replace-plus-pr444' into sortedrecords-zonemd-remove-replace

ximon18 committed:
Enhanced zone signing. (#418)

Co-authored-by: Jannik Peters <jannik.peters@posteo.de>

ximon18 committed:
Merge branch 'multiple-key-signing' into initial-nsec3-generation

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Correct outdated code comment.

ximon18 committed:
Improved/additional logging during NSEC3 generation.

ximon18 committed:
Remove commented out code.

ximon18 committed:
Make signing work with any objects as keys as long as they can answer the basic question are they to be used for signing keys or other zone content, making DnssecSigningKey one possible way to represent the key, not the only possible way.

ximon18 committed:
Minor import cleanup.

ximon18 committed:
Comment tweaks.

ximon18 committed:
FIX: Neither NSEC and NSEC3 nor hashing should include non-authoritative RR types in the Type Bitmap.

ximon18 committed:
Add Rtype::is_pseudo() for use by NSEC and NSEC3 logic.

ximon18 committed:
Implement MUST constraints from RFC 4034 and RFC 5155 excluding "pseudo" RRtypes from NSEC(3) type bitmaps.

ximon18 committed:
Replace the Signer with access to signing via new traits SignableZone and SignableZoneInPlace, removing the need for state to be held, especially as some of that state was better associated with the zone being signed than with the signer.

Introduce a new structure under sign/ to hold types extracted by refactoring the long records.rs into separate modules.

Add generalized RecordSet::update_data() instead of the very specific replace_soa() and replace_rrsign_for_apex_zonemd() functions.

ximon18 committed:
Cargo fmt.

ximon18 committed:
Fix doc tests.

ximon18 committed:
RustDoc fix.

ximon18 committed:
Better generic type name.

ximon18 committed:
More descriptive and consistent fn name.

ximon18 committed:
Add sorted_records::as_slice().

ximon18 committed:
Also allow RRsets to be signed via trait fn which is simpler than callers having to call generate_rrsigs() manually as it lacks the unnecessary Sorter type and add dnskeys param, and is more consistent with how signing of entire zones is now possible via trait too.

ximon18 committed:
Clippy.

ximon18 committed:
FIX: Add missing required dependency to fix broken compilation of the keyset example.

ximon18 committed:
Take out references to BIND and LDNS.

ximon18 committed:
De-duplicate SignableZone::sign_zone() and SignableZoneInPlace::sign_zone() by introducing SignableZoneInOut and RecordSlice.

ximon18 committed:
Remove the confusnig OctsMut generic type.

ximon18 committed:
Default TTL for newly created non-NSEC(3) RRs should be that of the SOA TTL.

ximon18 committed:
Add TODO comment.

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Remove unnecessary function.

ximon18 committed:
Use Deref instead of adding a new RecordSlice trait.

ximon18 committed:
Make it possible to construct SortedRecords without specifying the sorter, via Default.

ximon18 committed:
Make the Default SigningConfig actually have default behaviour.

ximon18 committed:
FIX: Don't panic when signing a zone that lacks a SOA.

ximon18 committed:
Start updating the RustDoc for the sign module.

ximon18 committed:
RustDoc formatting.

ximon18 committed:
Remove errant backticks in RustDoc.

ximon18 committed:
Use user supplied sort impl everywhere, and require CanonicalOrd.

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Group and move things around in the sign module.

ximon18 committed:
Fix doc tests.

ximon18 committed:
Use the generic parameter name Inner everywhere for consistency.

Replace &dyn DesignatedSigningKey with <DSK: DesignatedSigningKey> because the dyn doesn't gain anything as the underlying SigningKey takes a fixed Inner type.

ximon18 committed:
Consistency.

ximon18 committed:
Cargo fmt.

ximon18 committed:
Remove FamilyName, rename Family to OwnerRrs, and remove class checks, because all RRs in a zone per RFC 1035 should have the same class.

ximon18 committed:
As zone signing assumes, but does not check, that the zone is ordered, add a check in debug builds (not in release builds as it is too costly) if the zone is correctly sorted before signing.

ximon18 committed:
Revert "As zone signing assumes, but does not check, that the zone is ordered, add a check in debug builds (not in release builds as it is too costly) if the zone is correctly sorted before signing."

This reverts commit b1f7a20df54fe1f2f3c33cc9c926adc033dd30a9.

ximon18 committed:
Fix doc test.

ximon18 committed:
Pass an is_ent flag to the Nsec3Provider to allow it to be recorded for diagnostics such as those produced by ldns-signzone.

ximon18 committed:
Rename remaining references to family.

ximon18 committed:
Clippy.

ximon18 committed:
Cleanup:
- Remove unnecessary apex argument and trait fn as it can be derived from the input.
- Remove unnecesarsy skip_before, the input should be the zone and nothing else.
- Replace one unwrap() with an error return instead.
- The Sort generic argument should be called Sort everywhere, not S.
- SignableZone and SignableZoneInPlace don't no longer need specific impls for SortedRecords and Vec but can instead be blanket impls.
- Handle the case that the zone has too many apex SOAs.
- Factor out apex SOA lookup code.
- Added some RustDoc.

ximon18 committed:
RustDoc tweaks.

ximon18 committed:
RustDoc tweaks.

ximon18 committed:
RustDoc tweaks.

ximon18 committed:
RustDoc tweaks.

ximon18 committed:
FIX: When signing to another  collection rather than in-place don't neglect to sign the NSEC(3)s.

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Undo unintended changes compared to main.

ximon18 committed:
More RustDoc tweaks for the sign module, and restore the crypto common module by moving keys/keypair.rs to crypto/common.rs.

ximon18 committed:
Fix broken doc tests.

ximon18 committed:
Cargo fmt.

ximon18 committed:
Move crypto errors in to the main error submodule of sign.

ximon18 committed:
Re-export some types that live only in modules by the same name or whose submodule is useful for code structure but not for documentation.

ximon18 committed:
Fix missing feature guards.

ximon18 committed:
Ensure re-exports refer only to descendants of the current module.

ximon18 committed:
Add missing RRSIG term in RustDoc comment.

ximon18 committed:
Rename the hashing module to authnext (authenticated non-existence) as NSEC doesn't do any hashing, only NSEC3 does.

ximon18 committed:
Minor RustDoc tweaks.

ximon18 committed:
Cargo fmt.

ximon18 committed:
Rename the authnext module to authnonext which doesn't sound like the word "next".

ximon18 committed:
Rename authnonext to denial as ext is not really a good abbreviation of non-existence, and the full term is authenticated denial of existence.

ximon18 committed:
Move SigningConfig and signing::traits to the top of the sign module as they relate to signing in general, and rename the signing sub-module as it is specific to signature generation.

ximon18 committed:
Delete empty sign::zone sub-module.

ximon18 committed:
Minor RustDoc tweak.

ximon18 committed:
Add RustDoc for the `sign_zone()` function.

ximon18 committed:
More `sign_zone()` RustDoc.

ximon18 committed:
Typo correction.

ximon18 committed:
Typo correction.

ximon18 committed:
RustDoc correction.

ximon18 committed:
Cargo fmt.

ximon18 committed:
More RustDoc.

ximon18 committed:
Log each signed RRSET at trace level, not debug level.

ximon18 committed:
Remove unnecessary owner_name argument from sign_rrset(), don't require all callers to supply a scratch buffer by making sign_rrset() a thin wrapper around new sign_rrset_in() which has the code that was previously sign_rrset(), and add RustDoc.

ximon18 committed:
Reject attempts to sign an RRSIG RRset. (a) they should never be signed, and (b) they should not form RRsets.

ximon18 committed:
Reject invalid signature validity periods in sign_rrset().

ximon18 committed:
Typo fix in error message.

ximon18 committed:
Add a TODO comment.

ximon18 committed:
Add a debug time assert in sign_rrset() checking the label counts per RFC 4034.

ximon18 committed:
Add some RFC 4035 and 4035 based tests of sign_rrset().

ximon18 committed:
RustDoc updates for the sign module.

ximon18 committed:
Cargo fmt.

ximon18 committed:
Clippy.

ximon18 committed:
Fix messed up test code.

ximon18 committed:
Review feedback.

ximon18 committed:
Corrected a RustDoc comment.

ximon18 committed:
Corrected a RustDoc comment.

ximon18 committed:
Replace incorrect references to hashing which is only true for NSEC3, not for NSEC.

ximon18 committed:
Report the invalid signature validity period when sign_rrset() fails.

ximon18 committed:
No need to check for pseudo RTYPEs being added as the input ZoneRecordData type cannot represent them.

ximon18 committed:
Determine the TTL for NSEC records within generate_nsecs() because it is determined by the rules defined in RFC 9077 based on the apex SOA record. Return an error if no or multiple SOAs are found.

ximon18 committed:
Add RustDoc for generate_nsecs().

ximon18 committed:
Add tests for generate_nsecs().

ximon18 committed:
Cargo fmt.

ximon18 committed:
Clippy.

ximon18 committed:
Fix broken/missing RustDoc links.

ximon18 committed:
Minor test name corrections.

ximon18 committed:
- Make generate_rrsigs() take a config object instead of multiple config arguments, and make the apex optional as it isn't needed for a full zone.
- Split generate_rrsigs() apex handling functionality out to generate_apex_rrsigs().
- Split generate_rrsigs() key logging functionality out to log_keys_in_use().
- Use log::log_enabled() now that PR #465 has been merged.
- Add some initial generate_rrsigs() tests.
- Fix missing unwrap()s after calling .insert() in RustDoc example code.

ximon18 committed:
Minor cleanup of the way test keys are generated and used by generate_rrsig() tests.

ximon18 committed:
Require a version of Bytes that supports From<Box<[u8]>> (as Dnskey uses Box<[u8]> internally).

ximon18 committed:
FIX: Don't sign the apex twice.

ximon18 committed:
FIX: Don't skip signing when the apex isn't matched.

ximon18 committed:
- Move test helper functions to a shared module.
- Add a full zone test of generate_rrsigs() with default config.
- Change SigningKeyUsageStrategy to return Vec and have generate_rrsigs() sort and dedup the results of its fns so that DNSKEY RRs are generated in a deterministic order and that implementers don't have to create short lived HashSets. (even creating Vec isn't ideal...)

ximon18 committed:
Use SmallVec instead of Vec, to avoid allocation for a small temporary collection.

ximon18 committed:
And missing line break.

ximon18 committed:
Fix compilation error.

ximon18 committed:
FIX: At least one key for both roles is needed for signing.

ximon18 committed:
Additional RustDoc comments.

ximon18 committed:
FIX: Doc tests broken by recent logic fix.

ximon18 committed:
Default to adding missing DNSKEY RRs, as RFC 4035 section 2.1 requires it (SHOULD).

ximon18 committed:
FIX: Adding records to SortedRecords via iterator should also use extend, otherwise inserts via a large iterator will sort per insert which is slow comparing to dedup and sort after extend (and will make use of the Sorter too).

ximon18 committed:
Better parameter name.

ximon18 committed:
Minor improvements.

ximon18 committed:
Extend testing of generate_rrsigs() with a full zone to cover various key usage strategies.

ximon18 committed:
Clippy.

ximon18 committed:
FIX: Inverted flag.

ximon18 committed:
Organize imports.

ximon18 committed:
Add test for generating RRSIGs without adding DNSKEYs.

ximon18 committed:
Rename parameter.

ximon18 committed:
Use existing helper fns to simplify test code.

ximon18 committed:
Add missing [must_use] attributes.

ximon18 committed:
Correct / generalize old comments.

ximon18 committed:
Ah, the [must_use] are already inffered and duplicate and annoy Clippy, remove them again.

ximon18 committed:
Corrections and additions to the RustDoc for generate_rrsigs().

ximon18 committed:
Add a test of calling generate_rrsigs() on an already signed zone.

ximon18 committed:
- Remove the DNSKEY RRs from the input test zonefile as it is assumed to be unsigned.
- Add more tests and simplify some existing tests.

ximon18 committed:
- Also use SmallVec here.
- Add a test of generate_rrsigs() with mutliple KSKs and ZSKs.

ximon18 committed:
Merge branch 'main' into initial-nsec3-generation

ximon18 committed:
Normalize the generate_xxx interfaces to take config objects and return just the record types they produce.

ximon18 committed:
Fix broken doc test, restore flexible signature for Default impl for SortedRecords.

ximon18 committed:
More normalization of the generate_xxx interfaces to take config objects.

ximon18 committed:
Initial NSEC3 unit tests based on existing NSEC tests.

ximon18 committed:
Fix missing feature dependency.

ximon18 commented on src/sign/keys/signingkey.rs:
```diff
@@ -0,0 +1,152 @@
+    ///
+    /// The range spans from the inception timestamp up to and including the
+    /// expiration timestamp.
+    signature_validity_period: Option<RangeInclusive<Timestamp>>,
```
Hmm, yes, I see what you mean. This was fine for the initial use case of one-shot signing, but makes no sense for using the same key later to re-sign expired signatures. And mabe there are other reasons why it makes no sense. I'll change this.

ximon18 committed:
Take validity time for a signature as input to signing, not from a key, as the validity period required can change over the lifetime of a key and may be affected by other factors such as current signature expiration time and jitter.

ximon18 committed:
Impl Display for IntendedKeyPurpose.

Philip-NLnetLabs commented on src/sign/signatures/rrsigs.rs:
```diff
@@ -0,0 +1,1840 @@
+    N: ToName + Clone + Send,
+    D: RecordData
+        + ComposeRecordData
+        + From<Dnskey<Octs>>
```
From<Dnskey> is not needed and confusing

Philip-NLnetLabs commented on src/sign/signatures/rrsigs.rs:
```diff
@@ -0,0 +1,1840 @@
+    N: ToName + Clone + Send,
+    D: RecordData
+        + ComposeRecordData
+        + From<Dnskey<Octs>>
```
From\<Dnskey> is not needed and confusing

ximon18 replied:
Good catch. It was needed in an earlier version of the code but indeed is no longer needed.

ximon18 committed:
Review feedback: Remove From<Dnskey> as it is not needed.

ximon18 committed:
Remove unnecessary Send bounds.

ximon18 committed:
Add setter for RRSIg validity period to SigningConfig.

ximon18 committed:
Simplify NSEC unit test code.

ximon18 committed:
Simplify NSEC3 unit test code and fix up the occluded test copied from the NSEC unit test suite with the changes required for the NSEC3 case.

Philip-NLnetLabs commented on src/sign/signatures/rrsigs.rs:
```diff
@@ -0,0 +1,1840 @@
+where
+    Octs: AsRef<[u8]>,
+{
+    /// The NSEC3 records.
```
Wrong comment? Seems to be RRSIG records.

bal-e commented on src/sign/traits.rs:
```diff
@@ -0,0 +1,546 @@
+///
+/// [`sign_zone()`]: SignableZone::sign_zone
+pub trait SignableZone<N, Octs, Sort>:
+    Deref<Target = [Record<N, ZoneRecordData<Octs, N>>]>
```
This bound is incredibly restrictive -- it prevents the use of this trait with any zone representation beyond those already provided (e.g. an efficient zone representation where records are stored in a B-tree).  In order to be useful, these traits need to be loosened significantly, or should not be expressed as traits at all.

ximon18 replied:
Ideally there would be no reference to slices or arrays at all, only iterators. This evolved out of the existing code but isn't a desirable end state. Currently it is whatever compiles and works.

ximon18 committed:
- Added a unit test verifying that existing NSEC RRs are ignored by generate_nsecs().

- Added a unit test verifying compliance with the RFC 5155 Appendix A signed zone example.

- Imroved and added comments.

- Rewrote cryptic slice based NSEC3 next owner hashing loop with initial (untested!) type bit map merging and collision detection support, and removed no longer needed (but added in this branch) SortedRecords::as_mut_slice().

- Removed confusing duplicate storage of NSEC3 opt-out info in GenerateNsec3Config.

- Added missing (though not breaking) trailing periods in test data.

bal-e reviewed (commented):
This is a massive PR, @ximon18, well done on getting all this functionality together into a cohesive unit!

- My primary concern is with the large number of new traits added.  I'm worried that they're trying to provide extensibility in a manner that won't actually be helpful for our target use cases.  In some places, they can be replaced with simpler concrete data types, and I'd strongly prefer that.

- A lot of public methods (e.g. in `sign::records`) are still missing documentation.  Perhaps a Clippy lint can help catch everything?

- There are a number of similarly named types and traits (e.g. `DesignatedSigningKey`, `DnssecSigningKey`, and `SigningKey`).  Can some of these be combined together to reduce ambiguity in the public API?  I'm worried users will get lost in the number of types and traits they might have to think about.

ximon18 commented:
The missing docs are just not written yet.

ximon18 commented:
Re strategy traits, maybe. However, we use them in `dnst`. They exist at points where there is no one right way to do it, but a default is provided by us.

ximon18 commented:
Re similarly named types, yes it would be good to see how to improve this.

ximon18 commented on src/sign/signatures/strategy.rs:
```diff
@@ -0,0 +1,111 @@
+// to keep the response size for the DNSKEY query small, only keys designated
+// as being used to sign apex DNSKEY RRs (usually keys with the Secure Entry
+// Point (SEP) flag set) will be used to sign DNSKEY RRs.
+pub trait SigningKeyUsageStrategy<Octs, Inner>
+where
+    Octs: AsRef<[u8]>,
+    Inner: SignRaw,
+{
+    const NAME: &'static str;
+
+    fn select_signing_keys_for_rtype<
+        DSK: DesignatedSigningKey<Octs, Inner>,
+    >(
+        candidate_keys: &[DSK],
+        rtype: Option<Rtype>,
+    ) -> SmallVec<[usize; 4]> {
+        if matches!(rtype, Some(Rtype::DNSKEY)) {
+            Self::filter_keys(candidate_keys, |k| k.signs_keys())
+        } else {
+            Self::filter_keys(candidate_keys, |k| k.signs_zone_data())
+        }
+    }
+
+    fn filter_keys<DSK: DesignatedSigningKey<Octs, Inner>>(
+        candidate_keys: &[DSK],
+        filter: fn(&DSK) -> bool,
+    ) -> SmallVec<[usize; 4]> {
+        candidate_keys
+            .iter()
+            .enumerate()
+            .filter_map(|(i, k)| filter(k).then_some(i))
+            .collect()
+    }
+}
```
Which keys to use for what purpose is not defined by the keys themselves or even DNSKEY flag bits but instead by operator intent.

See `dnst` which implements several different strategies as different impls of this trait, to match behaviour offered by `ldns-signzone`.

ximon18 commented on src/sign/records.rs:
```diff
@@ -1,33 +1,96 @@
+/// Implement this trait to use a different sorting algorithm than that
+/// implemented by [`DefaultSorter`], e.g. to use system resources in a
+/// different way when sorting.
+pub trait Sorter {
+    /// Sort the given DNS resource records.
+    ///
+    /// The imposed order should be compatible with the ordering defined by
+    /// RFC 8976 section 3.3.1, i.e. _"DNSSEC's canonical on-the-wire RR
+    /// format (without name compression) and ordering as specified in
+    /// Sections 6.1, 6.2, and 6.3 of [RFC4034] with the additional provision
+    /// that RRsets having the same owner name MUST be numerically ordered, in
+    /// ascending order, by their numeric RR TYPE"_.
+    fn sort_by<N, D, F>(records: &mut Vec<Record<N, D>>, compare: F)
+    where
+        F: Fn(&Record<N, D>, &Record<N, D>) -> Ordering + Sync,
+        Record<N, D>: CanonicalOrd + Send;
+}
```
I was trying to avoid adding a rayon dependency to domain.

Testing with `dnst` showed huge speed up by using rayon to parallelize the sort.

The intent was not to split dingle vs multi-threaded but to support bring your own sort impl, e.g. rayon as one approach.

Vec was already in use by the existing code, plus I thought maybe for some users/use cases a vec is sufficient - who are we to say no you can't use a vec?

ximon18 commented on src/sign/records.rs:
```diff
@@ -45,274 +108,274 @@ impl<N, D> SortedRecords<N, D> {
-        apex: &FamilyName<ApexName>,
-        ttl: Ttl,
-    ) -> Vec<Record<N, Nsec<Octets, N>>>
+    pub fn owner_rrs(&self) -> RecordsIter<N, D> {
```
This already existed like this. Ideally I'd like to get rid of these iterators as I find them very confusing. This isn't records() because it doesn't iterate over records, it iterates over groups of records having the same owner name.

ximon18 commented on src/sign/records.rs:
```diff
@@ -45,274 +108,274 @@ impl<N, D> SortedRecords<N, D> {
     {
-        self.rrsets().find(|rrset| rrset.rtype() == Rtype::SOA)
-    }
+        let mut found_one = false;
+        loop {
+            if self.remove_first_by_name_class_rtype(
+                name.clone(),
+                class,
+                rtype,
+            ) {
+                found_one = true
+            } else {
+                break;
+            }
+        }
```
The ZONEMD support was added by @mozzieongit and incorporated as-is to this PR. I didn't get round to looking at it yet.

mozzieongit replied:
True, for a general task like the name suggests, this is terrible. I only used it to remove one or at most two records (ZONEMDs), and then it doesn't really matter, so I didn't think of improving it.

ximon18 commented on src/sign/records.rs:
```diff
@@ -45,274 +108,274 @@ impl<N, D> SortedRecords<N, D> {
+    pub fn remove_all_by_name_class_rtype(
+        &mut self,
+        name: N,
+        class: Option<Class>,
```
> A `SortedRecords` usually represents the records of a single zone, right?  Do we want to allow it to contain records of different classes?

The original code allowed for an arbitrary sequence of records potentially from multiple zones. I'm not sure why. I think supporting a single zone only makes more sense.

ximon18 commented on src/sign/records.rs:
```diff
@@ -1,33 +1,96 @@
+/// [`Sorter`].
 #[derive(Clone)]
-pub struct SortedRecords<N, D> {
+pub struct SortedRecords<N, D, Sort = DefaultSorter>
```
That's a point for @partim as the original author.

I' not very confirtable wuth the SortedRecords type to be honest, but it already existed and my approach was to evolve it not replace it, and I hoped later to have a deep enough understanding to revisit the whole design. 

I'd rather we had some way of saying we operate on types chosen by the user so long as they meet the necessary criteria. I'm not sure how to express unique and sorted per DNSSEC canonical order via bounds however.

ximon18 commented on src/sign/records.rs:
```diff
@@ -45,274 +108,274 @@ impl<N, D> SortedRecords<N, D> {
-pub struct Family<'a, N, D> {
+/// A set of records with the same owner name.
+#[derive(Clone)]
+pub struct OwnerRrs<'a, N, D> {
```
I'm not a fan of either, but family tells you nothing IMO, what is a "family" in FNS record terms? At least "owner" says something about what they have in common.

ximon18 commented on src/sign/records.rs:
```diff
@@ -470,33 +449,47 @@ impl<'a, N, D> Rrset<'a, N, D> {
     pub fn iter(&self) -> slice::Iter<'a, Record<N, D>> {
         self.slice.iter()
     }
+
+    pub fn len(&self) -> usize {
+        self.slice.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.slice.is_empty()
+    }
+
+    pub fn as_slice(&self) -> &'a [Record<N, D>] {
+        self.slice
+    }
+
+    pub fn into_inner(self) -> &'a [Record<N, D>] {
+        self.slice
+    }
```
I assume the intent was to deliberately restrict the interface to offer only meaningful operations for the intended context/use case.

ximon18 commented on src/sign/denial/nsec3.rs:
```diff
@@ -0,0 +1,1151 @@
+
+//------------ Nsec3HashProvider ---------------------------------------------
+
+pub trait Nsec3HashProvider<N, Octs> {
+    fn get_or_create(
+        &mut self,
+        apex_owner: &N,
+        unhashed_owner_name: &N,
+        unhashed_owner_name_is_ent: bool,
+    ) -> Result<N, Nsec3HashError>;
+}
```
> What is the purpose of this trait?  Can you explain cases where users can provide sufficiently different implementations?

`dnst` uses it to implement a capturing hash provider to allow it to dump diagnostic comments in the output zone. Normal usage doesn't need or want that overhead however. Prior to this trait it was a boolean flag with ghe capturing code miced in with the actual logic which made it harder to read, imposed a specific solution on the caller and added yet another flag to the input and an Option output that would not be used in normal usage. I thought this was a better approach.

ximon18 commented on src/sign/error.rs:
```diff
@@ -0,0 +1,270 @@
+/// from such an error differently (e.g. by foregoing signatures or informing
+/// an operator).
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct SignError;
```
> Should this be renamed to avoid confusion with `SigningError`?

Probably. I wanted some feedback on the design before bikeshedding names.

ximon18 commented on src/sign/keys/keymeta.rs:
```diff
@@ -0,0 +1,247 @@
+    /// and Zone Key flags on the key (i.e. whether the key is a KSK or ZSK or
+    /// something else).
+    purpose: IntendedKeyPurpose,
+
+    _phantom: PhantomData<(Octs, Inner)>,
```
Ok. If so it might be a leftover from refactoring.

ximon18 commented on src/sign/keys/keymeta.rs:
```diff
@@ -0,0 +1,247 @@
+
+//------------ DesignatedSigningKey ------------------------------------------
+
+pub trait DesignatedSigningKey<Octs, Inner>
+where
+    Octs: AsRef<[u8]>,
+    Inner: SignRaw,
+{
+    /// Should this key be used to "sign one or more other authentication keys
+    /// for a given zone" (RFC 4033 section 2 "Key Signing Key (KSK)").
+    fn signs_keys(&self) -> bool;
+
+    /// Should this key be used to "sign a zone" (RFC 4033 section 2 "Zone
+    /// Signing Key (ZSK)").
+    fn signs_zone_data(&self) -> bool;
+
+    fn signing_key(&self) -> &SigningKey<Octs, Inner>;
+}
```
`DnssecSigningKey` is one way to associate a key with intent, but the code using it doesn't care how it is structured, only that it can answer the questions posed by the trait. I didn't want to force this one way of doing it on the user.

ximon18 commented on src/sign/config.rs:
```diff
@@ -0,0 +1,103 @@
+
+    pub rrsig_validity_period_strategy: ValidityStrat,
+
+    _phantom: PhantomData<(Inner, KeyStrat, Sort)>,
```
I think I tried that and it didn't compile.

ximon18 commented on src/sign/denial/nsec3.rs:
```diff
@@ -0,0 +1,1151 @@
+}
+
+impl Nsec3ParamTtlMode {
+    pub fn fixed(ttl: Ttl) -> Self {
+        Self::Fixed(ttl)
+    }
+
+    pub fn soa() -> Self {
+        Self::Soa
+    }
+
+    pub fn soa_minimum() -> Self {
+        Self::SoaMinimum
+    }
```
No they are not.

The RFCs don't state how the TTL of the NSEC3PARAM RR should be chosen and different tools do it at least these three different ways.

ximon18 commented on src/base/iana/zonemd.rs:
```diff
@@ -0,0 +1,50 @@
+    /// the [ZONEMD].
+    ///
+    /// For the currently registered values see the [IANA registration]. This
+    /// type is complete as of 2024-11-29.
```
I think this was copied from existing `domain` code, i.e. aimed to be consistent.

ximon18 commented on src/base/iana/zonemd.rs:
```diff
@@ -0,0 +1,50 @@
+    =>
+    ZonemdAlg, u8;
+
+    /// Specifies that the SHA-384 algorithm is used.
+    (SHA384 => 1, "SHA384")
+
+    /// Specifies that the SHA-512 algorithm is used.
+    (SHA512 => 2, "SHA512")
```
I think this was copied from existing `domain` code, i.e. aimed to be consistent.

ximon18 commented on src/rdata/nsec3.rs:
```diff
@@ -496,6 +507,35 @@ impl<Octs> Nsec3param<Octs> {
 
+//--- Default
+
+impl<Octs> Default for Nsec3param<Octs>
+where
+    Octs: From<&'static [u8]>,
+{
+    /// Best practice default values for NSEC3 hashing.
+    ///
+    /// Per [RFC 9276] section 3.1:
+    ///
+    /// - _SHA-1, no extra iterations, empty salt._
+    ///
+    /// Per [RFC 5155] section 4.1.2:
+    ///
+    /// - _The Opt-Out flag is not used and is set to zero._
+    /// - _All other flags are reserved for future use, and must be zero._
+    ///
+    /// [RFC 5155]: https://www.rfc-editor.org/rfc/rfc5155.html
+    /// [RFC 9276]: https://www.rfc-editor.org/rfc/rfc9276.html
+    fn default() -> Self {
+        Self {
+            hash_algorithm: Nsec3HashAlg::SHA1,
+            flags: 0,
+            iterations: 0,
+            salt: Nsec3Salt::empty(),
+        }
+    }
+}
```
I think you make a good point.

My intent was to provide good defaults.

ximon18 commented on src/rdata/zonemd.rs:
```diff
@@ -233,20 +233,7 @@ impl<Octs: AsRef<[u8]>> ZonefileFmt for Zonemd<Octs> {
         p.block(|p| {
             p.write_token(self.serial)?;
             p.write_show(self.scheme)?;
-            p.write_comment(format_args!("scheme ({})", match self.scheme {
-                Scheme::Reserved => "reserved",
-                Scheme::Simple => "simple",
-                Scheme::Unassigned(_) => "unassigned",
-                Scheme::Private(_) => "private",
-            }))?;
             p.write_show(self.algo)?;
-            p.write_comment(format_args!("algorithm ({})", match self.algo {
-                Algorithm::Reserved => "reserved",
-                Algorithm::Sha384 => "SHA384",
-                Algorithm::Sha512 => "SHA512",
-                Algorithm::Unassigned(_) => "unassigned",
-                Algorithm::Private(_) => "private",
-            }))?;
```
@mozzieongit ?

mozzieongit replied:
I don't think I touched that. Did you remove that while moving the ZONEMD types to the iana base module?

ximon18 replied:
Hmm. Will take a closer look.

ximon18 commented on src/sign/config.rs:
```diff
@@ -0,0 +1,103 @@
+//------------ SigningConfig -------------------------------------------------
+
+/// Signing configuration for a DNSSEC signed zone.
+pub struct SigningConfig<
+    N,
+    Octs,
+    Inner,
+    KeyStrat,
+    ValidityStrat,
+    Sort,
+    HP = OnDemandNsec3HashProvider<Octs>,
+> where
```
Have you seen the openid connect and oauth2 crates? 🤪

I'm honestly not sure. To me they offer choice and control, but they are a bit of a rough edge to work with in Rust.

I'd be happy with an alternative that meets our needs, this arose out of meeting the needs of `ldns-signzone` in a way that didn't hardcode choices for `ldns-signzone` into `dnst`, and assumes other users may similarly need/want that control from a library.

ximon18 commented on src/sign/config.rs:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    pub fn set_rrsig_validity_period_strategy(
+        &mut self,
+        rrsig_validity_period_strategy: ValidityStrat,
+    ) {
+        self.rrsig_validity_period_strategy = rrsig_validity_period_strategy;
+    }
```
The latter I think.

ximon18 committed:
FIX: Inverted flag.

ximon18 committed:
Review feedback: Wrong comment.

ximon18 committed:
FIX: Use the supplied sorter.

ximon18 committed:
- Remove NSEC3 type bit map merging as it is not necessary due to the requirement that the input be sorted.
- Return an error on collision.
- Add a collision test.
- More unwrap -> Err.
- More comments.

ximon18 committed:
Clippy.

ximon18 committed:
Remove temporary init_logging() helper fn.

ximon18 committed:
Add a test for hashing not producing the expected result.

Philip-NLnetLabs committed:
Merge branch 'main' into initial-nsec3-generation

Philip-NLnetLabs committed:
The great rename plus cleanup. (#488)

* The great rename.

* Most of crypto requires ring.

* Introduce unstable-sign and make sign conditional on ring.

* Make cypto depend on unstable-sign for now.

* Get rid of the apex_owner parameter to sign_rrset_in. Take the name from the key.

* Fix tests.

* Remove NsecToNsec3TransitionState, Nsec3ToNsecTransitionState,
TransitioningNsecToNsec3, and TransitioningNsec3ToNsec.

* Remove RrsigValidityPeriodStrategy.

* Fix tests.

* No longer generate and sign the DNSKEY RRset.

* Fix documentation example.

* Introduce crypto::common::DigestContext, move nsec3_hash to dnssec::common.

* Add SHA-256 and use DigestContext in the signature cache.

* Move #[allow(unreachable_code)] to a different place.

* Move DnskeyExt from crypto::validate to dnssec::validator::base.

* Add crypto::common::PublicKey. Move RrsigExt to dnssec::validator::base

* Removed Key because it is mostly redundant.

Move parse_from_bind, format_as_bind, and display_as_bind to dnssec::common.
Move key_size to dnssec::validator::base.

* Remove PublicKeyBytes.

* Bump ring version to 0.17.2 to get AsRef<[u8]> for UnparsedPublicKey.

* Clippy.

* Fmt

* crypto is now independent of dnssec::sign.

* Fix some feature test issues.

* More feature issues.

* Remove hard dependency on ring for DNSSEC signing and validation.

* Fix dependencies for the keyset example.

* Introduce crypto::sign, get rid of crypto::misc.

* Improve documentation for the crypto module.

* Clippy found places where documentation was missing.

* Cleanup documentation for the dnssec module.

* Spell feature flags correctly.

* Improve feature tests.

* Remove Nsec3HashProvider.

* Fix example.

* Rename Default to Rsa.

* Add support for ED448 to from_dnskey.

* Add support for MessageDigest::sha1 and MessageDigest::sha512 to dnskey.

* Use unreachable for some match cases that cannot happen.

* Some simplifications.

* Spelling correction.

* Typo correction.

* Update src/dnssec/sign/denial/config.rs

Co-authored-by: Ximon Eighteen <3304436+ximon18@users.noreply.github.com>

* Update Cargo.toml

Co-authored-by: Ximon Eighteen <3304436+ximon18@users.noreply.github.com>

* Remove previous unstable-sign feature line.

* Restore stelline.

* Small doc fix.

* remove superfluous map_err.

* Rename DigestContext to DigestBuilder.

* Also rename in doc and test.

* Move doc string.

* Use SecAlg instead of pointers.

* Extract common RSA encoding code.

* Some comments.

* Move tests around.

* [crypto] Make minor changes

Following PR comments (https://github.com/NLnetLabs/domain/pull/488).

---------

Co-authored-by: Ximon Eighteen <3304436+ximon18@users.noreply.github.com>
Co-authored-by: arya dradjica <arya@nlnetlabs.nl>

Philip-NLnetLabs committed:
Merge branch 'main' into initial-nsec3-generation

Philip-NLnetLabs committed:
Fix spelling of no_run.
